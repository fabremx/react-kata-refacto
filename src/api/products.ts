import { Product } from "../models";

export const getProducts = async (): Promise<Product[]> => {
  const response = await fetch(`http://localhost:3000/api/products.json`);
  const { products }: { products: Product[] } = await response.json();
  return products;
};
