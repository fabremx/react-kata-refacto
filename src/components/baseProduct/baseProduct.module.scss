@import "../../assets/styles/colors.scss";

td,
th {
  text-align: left;
  padding: 0.75rem;
}

td {
  border-bottom: 1px solid #dee2e6;
}

td.photo {
  width: 120px;
  height: 110px;
  text-align: center;

  img {
    max-width: 100%;
    max-height: 85px;
  }
}

td.name span,
td.photo span {
  font-size: 12px;
  margin-left: 10px;
}

td.quantity input {
  border: 2px solid #000000;
  text-align: center;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  /* Firefox */
  &[type="number"] {
    -moz-appearance: textfield;
  }
}

td.remove {
  text-align: center;

  svg {
    width: 14px;
    fill: $text-secondary-lb;

    &:hover {
      cursor: pointer;
      fill: $text-secondary-db;
    }
  }
}
