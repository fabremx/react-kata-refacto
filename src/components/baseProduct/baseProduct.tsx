import { ReactElement } from "react";
import { ReactComponent as RemoveIcon } from "../../assets/images/remove.svg";
import { Product } from "../../models";
import styles from "./baseProduct.module.scss";

interface Props {
  product: Product;
  children: ReactElement;
  removeProduct: (id: number) => void;
}

export function BaseProduct({
  product,
  children,
  removeProduct,
}: Props): ReactElement {
  return (
    <tr key={product.id}>
      {children}

      <td className={styles.quantity}>
        <input
          type="number"
          size={4}
          defaultValue={product.quantity}
          min="0"
          step="1"
          disabled
        />
      </td>

      <td className={styles.total}>$ {product.price * product.quantity}</td>

      <td className={styles.remove}>
        <RemoveIcon
          onClick={() => removeProduct(product.id)}
          data-testid="remove"
        />
      </td>
    </tr>
  );
}
