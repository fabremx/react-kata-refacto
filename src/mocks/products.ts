import { Product } from "../models";

const BASE_PRODUCT = {
  price: 10,
  image: "image 1",
  quantity: 2,
};

export const NEW_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 1,
  title: "New Product",
  isNew: true,
  isSoonEnding: false,
};
export const ENDED_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 2,
  title: "Ended Product",
  isNew: false,
  isSoonEnding: true,
};
export const COMMON_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 3,
  title: "Common Product",
  isNew: false,
  isSoonEnding: false,
};
