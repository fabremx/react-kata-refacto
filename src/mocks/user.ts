import { User } from "../models";

const BASE_USER = {
  id: 1,
  firstname: "firstname",
  lastname: "lastname",
};
export const MOCKED_USER = {
  ...BASE_USER,
  isVIP: false,
  coupon: 0,
};
export const USER_VIP_WITH_COUPON: User = {
  ...BASE_USER,
  isVIP: true,
  coupon: 10,
};
export const USER_VIP_WITHOUT_COUPON: User = {
  ...BASE_USER,
  isVIP: true,
  coupon: 0,
};

export const USER_NOT_VIP_WITH_COUPON: User = {
  ...BASE_USER,
  isVIP: false,
  coupon: 10,
};
export const USER_NOT_VIP_WITHOUT_COUPON: User = {
  ...BASE_USER,
  isVIP: false,
  coupon: 0,
};
