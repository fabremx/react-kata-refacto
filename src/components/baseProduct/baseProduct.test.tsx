import { render } from "@testing-library/react";
import { COMMON_PRODUCT } from "../../mocks/products";
import { BaseProduct } from "./baseProduct";

jest.mock("../productPrice", () => ({
  ProductPrice: () => <>Mock ProductPrice</>,
}));

describe("BaseProduct", () => {
  it("should render correctly the component", () => {
    // Given
    const tableRow = document.createElement("tbody");
    // When
    const { asFragment } = render(
      <BaseProduct
        product={COMMON_PRODUCT}
        removeProduct={jest.fn()}
        children={<td>Mock</td>}
      />,
      {
        container: document.body.appendChild(tableRow),
      }
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });
});
