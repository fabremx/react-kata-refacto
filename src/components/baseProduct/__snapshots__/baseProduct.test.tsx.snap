// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseProduct should render correctly the component 1`] = `
<DocumentFragment>
  Mock
  <input
    disabled=""
    min="0"
    size="4"
    step="1"
    type="number"
    value="2"
  />
  $ 20
  <svg
    data-testid="remove"
  >
    remove.svg
  </svg>
</DocumentFragment>
`;
